src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -688,6 +688,12 @@ public virtual void OnDeserialization(object? sender)
             int hashsize = siInfo.GetInt32(HashSizeName);
             _comparer = (IEqualityComparer<TKey>)siInfo.GetValue(ComparerName, typeof(IEqualityComparer<TKey>))!; // When serialized if comparer is null, we use the default.
 
+            // "==" would fail since they're different instances
+            if (Equals(_comparer, EqualityComparer<TKey>.Default))
```

Comment:
This can break if the custom equality comparer had non-standard or broken implemenation of Equals.

Also, this does not handle all cases of the default compsres. You are still going to see performance regression after roundtrip in some cases.

As I have said in the Issue, binary serialization is legacy (and very complex) feature that se do not wish to improve. I do not think we should be doping any improvements here.

Reply:
Understood. I agree this part isn't an important place to tackle with. Please feel free to close the PR if sees appropriate.